### PR TITLE
rustdoc: fix cross-crate `impl Sized` & `impl ?Sized`

### PR DESCRIPTION
**Pull request by fmease**
Previously, cross-crate impl-Trait (APIT, RPIT, etc.) that only consists of a single `Sized` bound (modulo outlives-bounds) and ones that are `?Sized` were incorrectly rendered. To give you a taste (before vs. after):

```diff
- fn sized(x: impl ) -> impl 
+ fn sized(x: impl Sized) -> impl Sized

- fn sized_outlives<'a>(x: impl 'a) -> impl 'a
+ fn sized_outlives<'a>(x: impl Sized + 'a) -> impl Sized + 'a

- fn maybe_sized(x: &impl ) -> &impl 
+ fn maybe_sized(x: &impl ?Sized) -> &impl ?Sized

- fn debug_maybe_sized(x: &impl Debug) -> &impl ?Sized + Debug
+ fn debug_maybe_sized(x: &(impl Debug + ?Sized)) -> &(impl Debug + ?Sized)
```

Moreover, we now surround impl-Trait that has multiple bounds with parentheses if they're the pointee of a reference or raw pointer type. This affects both local and cross-crate docs. The current output isn't correct (rustc would emit the error *ambiguous `+` in a type* if we fed the rendered code back to it).

---

Best reviewed commit by commit :)

@rustbot label A-cross-crate-reexports

### CODE REVIEW

File: src/librustdoc/clean/mod.rs [hunk 1]
```diff
@@ -826,33 +826,26 @@ fn clean_ty_generics<'tcx>(
...
-                    p.get_bounds()
+                bounds.extend(
+                    pred.get_bounds()
                         .into_iter()
                         .flatten()
                         .cloned()
-                        .filter(|b| !b.is_sized_bound(cx)),
                 );
 
-                let proj = projection.map(|p| {
```
Comment (fmease) [inline src/librustdoc/clean/mod.rs:841]: This is just a refactor. I found the `map` followed by `and_then`, try-operator and if-let quite hairy.

File: src/librustdoc/clean/mod.rs
Comment (fmease) [inline src/librustdoc/clean/mod.rs:1]: This further extends the preexisting bound cleaning duplication of `clean_ty_generics` and `clean_middle_opaque_bounds`. Let me know if this is okay for the time being, I plan on consolidating these implementations next when I overhaul bound cleaning to fix #113015.

**Comment by fmease** [comment]
I can try to split the commit into two to disentangle semantic changes from stylistic ones if you find the diff too illegible.

File: src/librustdoc/clean/mod.rs [hunk 2]
```diff
@@ -862,8 +861,27 @@ fn clean_ty_generics<'tcx>(
         .collect::<Vec<_>>();
 
     for (param, mut bounds) in impl_trait {
+        let mut has_sized = false;
+        bounds.retain(|b| {
+            if b.is_sized_bound(cx) {
+                has_sized = true;
```
Comment (fmease) [inline src/librustdoc/clean/mod.rs:867]: Looking at this, I don't think it handles `~const Sized` (`TraitBoundModifier::MaybeConst`) correctly.
I pretty sure it results in `has_sized == false` making us add `+ ?Sized` to it.
Let me know if I should fix it here or in a follow-up (I guess there might be some more misuses of `is_sized_bound`).

Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs:867]: That answers the question then. ;)

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors r+ rollup

**Pull request merged**